### PR TITLE
maxdatapoints

### LINKED ISSUE #1957: Handle maxDataNodes and maxDataPoints
When attempting to copy log data from a server with larger values of maxDataNodes or maxDataPoints, the copy operation will fail due to attempting to send more data than the target server allows. Handle this by reworking all workers that handle log data to take maxDataNodes or maxDataPoints into consideration by splitting Add/Update queries into smaller chunks if they are larger than what the target server allows.

### PR DESCRIPTION
**Pull request by robertbasti**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

[//]: # (Write the GitHub issue number starting with #, or, for Equinor only, the Jira issue number starting with WE-)

This pull request fixes #1957 

## Description

* checks  maxDataNodes or maxDataPoints of the server to get maximum capacity and splits queries into chunks

[//]: # (Please include a written summary of the changes and which issue is fixed. List any dependencies that are required for this change._)

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [ ] New feature (non-breaking change which adds functionality)
* [X] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [ ] Frontend
* [X] API
* [ ] WITSML
* [ ] Desktop
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [x] New code is covered by passing tests

### CODE REVIEW

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs [hunk 1]
```diff
@@ -59,6 +67,56 @@ protected string CancellationReason()
...
+                            Data = chunk.ToList(),
+                        }
+                    }.AsItemInList()
+                }
+            ).ToList();
+
+            return batchedQueries;
+        }
+
+        protected async Task<int> GetMaxBatchSize(List<string> mnemonics, string functionType, string queryTypeName)
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: If this is to be used by anything other than logs, it makes more sense to accept the number of columns instead of the list of mnemonics.

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs [hunk 2]
```diff
@@ -59,6 +67,56 @@ protected string CancellationReason()
...
+        {
+            var targetServerCapabilities = await GetTargetServerCapabilities();
+            var serverCapabilites =
+                targetServerCapabilities.ServerCapabilities;
+
+            var functions = serverCapabilites.Select(x => x.Functions.Find(y => y.Name.Equals(functionType)));
+            var logCapabilities = functions.Select(x =>
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: rename to objectCapabilities

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs [hunk 3]
```diff
@@ -59,6 +67,56 @@ protected string CancellationReason()
...
+        protected async Task<WitsmlCapServers> GetTargetServerCapabilities()
+        {
+            if (_targetServerCapabilities == null)
+            {
+                _targetServerCapabilities = await GetTargetWitsmlClientOrThrow().GetCap();
+            }
+            return _targetServerCapabilities;
+        }
+
+        protected List<WitsmlLogs> GetUpdateLogDataQueries(WitsmlLog log, WitsmlLogData logData, int chunkSize, string mnemonicList)
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: This method is specific to logs, so it makes more sense to move it to another file. We have a LogWorkerTools file, maybe that's better?

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs [hunk 4]
```diff
@@ -438,5 +441,8 @@ private class CopyResult
             public int OriginalNumberOfRows { get; init; }
             public string ErrorReason { get; init; }
         }
+
+
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs]: Remove these extra lines

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs [hunk 5]
```diff
@@ -55,7 +55,24 @@ public override async Task<(WorkerResult, RefreshAction)> Execute(ImportLogDataJ
             }
 
             //Todo: find a way to determine the maximum amount of rows that can be sent to the WITSML server then pass that amount to the CreateImportQueries method
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs]: This comment can now be removed

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs [hunk 6]
```diff
@@ -59,6 +67,56 @@ protected string CancellationReason()
...
+                {
+                    Logs = new WitsmlLog
+                    {
+                        Uid = log.Uid,
+                        UidWell = log.UidWell,
+                        UidWellbore = log.UidWellbore,
+                        LogData = new WitsmlLogData
+                        {
+                            MnemonicList = mnemonicList,
+                            UnitList = logData.UnitList,
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: You take the unitList from logData, but the mnemonicList as an input parameter. I think we should do the same for both for consistency.

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs [hunk 7]
```diff
@@ -55,7 +55,24 @@ public override async Task<(WorkerResult, RefreshAction)> Execute(ImportLogDataJ
...
-            WitsmlLogs[] queries = CreateImportQueries(job, chunkSize).ToArray();
+            var ro = job.DataRows
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs]: Rename to something like _dataRows_ to make it easier to read than just _ro_.

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs:1]: Can you remove the CreateImportQueries function as it is not used anymore?

File: Src/WitsmlExplorer.Api/Workers/OffsetLogCurveWorker.cs [hunk 8]
```diff
@@ -176,7 +177,11 @@ private async Task DeleteLogData(WitsmlLog log, WitsmlLogCurveInfo logCurveInfo,
 
         private async Task UpdateLogData(WitsmlLog log, WitsmlLogCurveInfo logCurveinfo, WitsmlLogData offsetLogData)
         {
-            var queries = GetUpdateLogDataQueries(log, offsetLogData);
+            var mnemonics = log.LogCurveInfo.Select(lci => lci.Mnemonic).ToList();
```
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/OffsetLogCurveWorker.cs]: Can you use the mnemonicList below instead of using lci? The log can have more lci elements than what we actually send in a query, so using the list from the data should be more accurate.

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs (hunk 3 above)
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: You only use the WitsmlLog to get the uids. At some of the places where you use this function, you create a new WitsmlLog only to add the uids. In this case it might be better to accept the uids separately instead.

File: Tests/WitsmlExplorer.Api.Tests/Workers/CopyLogDataWorkerTests.cs [hunk 9]
```diff
@@ -244,6 +246,7 @@ public async Task Execute_DifferentIndexMnemonics_CopyWithTargetIndex()
             (WorkerResult result, RefreshAction _) = await _worker.Execute(job);
             Assert.True(result.IsSuccess);
             Assert.Equal(3, updatedLogs.First().Logs.First().LogCurveInfo.Count);
+            var aa = updatedLogs[1].Logs.First().LogData.MnemonicList;
```
Comment (eliasbruvik) [inline Tests/WitsmlExplorer.Api.Tests/Workers/CopyLogDataWorkerTests.cs]: Is this addition intentional?

**Review by eliasbruvik — CHANGES_REQUESTED** [review]
Good job, I like the generic implementation. I left a few comments for improvements.

File: Tests/WitsmlExplorer.Api.Tests/Workers/CopyLogDataWorkerTests.cs (hunk 9 above)
Comment (robertbasti) [reply Tests/WitsmlExplorer.Api.Tests/Workers/CopyLogDataWorkerTests.cs]: fixed

File: Src/WitsmlExplorer.Api/Workers/OffsetLogCurveWorker.cs (hunk 8 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/OffsetLogCurveWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs:1]: done

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs (hunk 7 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs (hunk 6 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: this is because in CopyLogDataWorker it is handled a little bit differently. In the solution you proposed, unit tests would fail.

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs (hunk 5 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs (hunk 4 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/Copy/CopyLogDataWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs (hunk 1 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs (hunk 2 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/BaseWorker.cs (hunk 3 above)
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: done

Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/BaseWorker.cs]: done

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs:1]: There is a chunkSize variable defined at the top of this file that is used in the logging. Remove the variable, and log the actual chunkSize.

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Workers/ImportLogDataWorker.cs:1]: fixed

**Review by eliasbruvik — APPROVED** [review]
*Reviewer checklist*
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code
